cmd/record/record.go

```diff
@@ -0,0 +1,217 @@
+	"google.golang.org/grpc"
+)
+
+const (
+	fitlerSep     = " "
+	resetLastLine = "\033[A\033[2K"
+)
+
+var fileSinkPrefix string
+var maxCaptureLength uint32
+
+// New creates a new record subcommand
+func New(vp *viper.Viper) *cobra.Command {
+	recordCmd := &cobra.Command{
+		Use:   "record [flags] filter1 filter2 ... filterN",
+		Short: "Capture and record network packets",
+		Long:  "Capture and record network packets into a pcap file stored on the Hubble server",
+		Example: `
+Record all TCP traffic from 192.168.1.0/24 (any source port) to 10.0.0.0/16 port 80:
+
+  hubble record "192.168.1.0/24 0 10.0.0.0/16 80 TCP"
+
+Multiple filters may be specified, in which case any matching packet will be recorded.
+The filter syntax is "srcPrefix srcPort dstPrefix dstPort proto". Currently supported
+protocols are TCP, UDP and ANY.`,
+		Hidden: true, // this command is experimental
+		RunE: func(cmd *cobra.Command, args []string) error {
+			filters, err := parseFilters(args)
+			if err != nil {
+				return err
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			hubbleConn, err := conn.New(ctx, vp.GetString(config.KeyServer), vp.GetDuration(config.KeyTimeout))
+			if err != nil {
+				return err
+			}
+			defer hubbleConn.Close()
+
+			return runRecord(ctx, hubbleConn, filters)
+		},
+	}
+
+	recorderFlags := pflag.NewFlagSet("recorder", pflag.ContinueOnError)
+	recorderFlags.StringVar(&fileSinkPrefix, "file-prefix", "hubble", "File prefix of the resulting pcap file")
+	recorderFlags.Uint32Var(&maxCaptureLength, "max-capture-len", 0, "Sets the maximum capture length (zero for full capture)")
+
+	recordCmd.Flags().AddFlagSet(recorderFlags)
+	recordCmd.SetUsageTemplate(template.Usage(config.ServerFlags, recorderFlags))
+
+	return recordCmd
+}
+
+func parseProto(proto string) (recorderpb.Protocol, error) {
+	switch strings.ToUpper(proto) {
+	case "TCP":
+		return recorderpb.Protocol_PROTOCOL_TCP, nil
+	case "UDP":
+		return recorderpb.Protocol_PROTOCOL_UDP, nil
+	case "ANY":
+		return recorderpb.Protocol_PROTOCOL_ANY, nil
+	default:
+		return 0, fmt.Errorf("unsupported protocol: %s", proto)
+	}
+}
+
+func parseFilters(filters []string) ([]*recorderpb.Filter, error) {
+	if len(filters) == 0 {
+		return nil, errors.New("need to specify at least one filter")
+	}
+
+	result := make([]*recorderpb.Filter, 0, len(filters))
+	for _, filter := range filters {
+		f := strings.SplitN(filter, fitlerSep, 5)
+		if len(f) != 5 {
+			return nil, fmt.Errorf("invalid 5-tuple filter: %q", f)
+		}
+
+		srcPrefix := f[0]
+		srcPort, err := strconv.ParseUint(f[1], 10, 16)
+		if err != nil {
+			return nil, fmt.Errorf("invalid source port: %s", f[1])
+		}
+		dstPrefix := f[2]
+		dstPort, err := strconv.ParseUint(f[3], 10, 16)
+		if err != nil {
+			return nil, fmt.Errorf("invalid destination port: %s", f[3])
+		}
+		proto, err := parseProto(f[4])
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, &recorderpb.Filter{
```

Comment:
is the plan to leave all the error handling for the server, or add some validation in the client side too?

Reply:
We also don't so client-side validation for almost all filters in the normal observe API, so I think it's fine to leave this server-side, also to avoid the two implementations diverging.